pkg/landscaper/blueprints/store.go

```diff
@@ -0,0 +1,303 @@
+	index    cache.Index
+	fs       vfs.FileSystem
+
+	size        int64
+	currentSize int64
+	// usage describes the actual usage of the filesystem.
+	// It calculated by using the max size and the current size.
+	usage float64
+
+	gcConfig      config.GarbageCollectionConfiguration
+	resetStopChan chan struct{}
+}
+
+// NewStore creates a new blueprint cache using a base filesystem.
+//
+// The caller should always close the cache for a graceful termination.
+//
+// The store should be initialized once as this is a global singleton.
+func NewStore(log logr.Logger, baseFs vfs.FileSystem, config config.BlueprintStore) (*Store, error) {
+	if log == nil {
+		log = logr.Discard()
+	}
+
+	if len(config.Path) != 0 {
+		var err error
+		config.Path, err = vfs.TempDir(baseFs, baseFs.FSTempDir(), "bsStore")
+		if err != nil {
+			return nil, fmt.Errorf("unable to setup temporary directory for blueprint store: %w", err)
+		}
+	}
+
+	fs, err := projectionfs.New(baseFs, config.Path)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create store filesystem: %w", err)
+	}
+
+	store := &Store{
+		log:      log,
+		disabled: config.DisableCache,
+		index:    cache.NewIndex(),
+		fs:       fs,
+		gcConfig: config.GarbageCollectionConfiguration,
+	}
+
+	if config.Size != "0" {
+		quantity, err := resource.ParseQuantity(config.Size)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse size %q: %w", config.Size, err)
+		}
+		sizeInBytes, ok := quantity.AsInt64()
+		if !ok {
+			return nil, fmt.Errorf("unable to parse size %q as int", config.Size)
+		}
+		store.size = sizeInBytes
+
+		store.StartResetInterval()
+	}
+	return store, nil
+}
+
+// DefaultStore creates a default blueprint store.
+func DefaultStore(fs vfs.FileSystem) (*Store, error) {
+	defaultStoreConfig := config.BlueprintStore{}
+	cs := v1alpha1.BlueprintStore{}
+	v1alpha1.SetDefaults_BlueprintStore(&cs)
+	if err := v1alpha1.Convert_v1alpha1_BlueprintStore_To_config_BlueprintStore(&cs, &defaultStoreConfig, nil); err != nil {
+		return nil, err
+	}
+	return NewStore(logr.Discard(), memoryfs.New(), defaultStoreConfig)
+}
+
+func (s *Store) Close() error {
+	close(s.resetStopChan)
+	return nil
+}
+
+// Store stores a blueprint on the given filesystem.
+// It is expected that the bpReader contains a tar archive.
+func (s *Store) Store(ctx context.Context, cd *cdv2.ComponentDescriptor, resource cdv2.Resource, bpReader io.Reader) (*Blueprint, error) {
+	bpID := blueprintID(cd, resource)
+	bpPath := blueprintPath(bpID)
+	if bp, err := s.Get(ctx, cd, resource); err == nil {
+		return bp, nil
+	}
+
+	if _, err := s.fs.Stat(bpPath); err == nil {
+		if err := s.fs.RemoveAll(bpPath); err != nil {
+			s.log.Error(err, "unable to cleanup directory")
+			return nil, errors.New("unable to cleanup store")
+		}
+	}
+
+	if err := s.fs.Mkdir(bpPath, os.ModePerm); err != nil && !os.IsExist(err) {
+		return nil, fmt.Errorf("unable to create bluprint directory: %w", err)
+	}
+	s.mux.Lock()
```

Comment:
Why does the mutex not include the s.fs cleanup at the beginning of the method?

Reply:
ah yes good point. The cleanup and `create dir` should be included in the mutex